### PR TITLE
[Prow CI workflow Enablement] Use MCE create HostedCluster on AWS + KubeVirt platform

### PR DESCRIPTION
**Pull request by LiangquanLi930**
The purpose of this PR is to build a ci workflow for installing the kubevirt hypershift cluster by MCE version 2.2.

Reference:
The MCE operator installation. https://github.com/stolostron/deploy/tree/master/multiclusterengine

Create a hyperShift cluster with MCE operator 2.2.
https://github.com/stolostron/hypershift-addon-operator/blob/main/docs/provision_hosted_cluster_on_mce_local_cluster.md
mce-aws
https://github.com/openshift/release/pull/33782

[OCPQE-13739](https://issues.redhat.com/browse/OCPQE-13739)

### CODE REVIEW

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

File: ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    - chain: hypershift-mce-kubevirt-destroy-2.2
+    - chain: ipi-aws-post
+    test:
+    - chain: hypershift-conformance
+    pre:
+    - chain: ipi-aws-ovn-pre
+    - chain: ipi-install
+    - ref: hypershift-kubevirt-install
+    - ref: hypershift-mce-install
+    - chain: hypershift-mce-kubevirt-create-2.2
```
Comment (davidvossel) [inline ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml:20]: Does this `hypershift-mce-kubevirt-create-2.2` step need to be tied directly to mce 2.2? what happens when we release and begin testing 2.3, 2.4, etc... Will we have to have a different set of steps for each mce version?

Is it possible that we simply have `hypershift-mce-kubevirt-create` and `hypershift-mce-kubevirt-destroy` steps which have the MCE_VERSION passed into them in order to account for any specific changes in behavior between the different mce versions in the future?

File: ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+workflow:
+  as: hypershift-mce-kubevirt-conformance
+  documentation: |-
+    The HyperShift mce kubevirt conformance workflow executes tests against a new ephemeral
+  steps:
+    env:
+      COMPUTE_NODE_TYPE: c5n.metal
+      CONTROL_PLANE_INSTANCE_TYPE: m4.xlarge
```
[the remark this answers is not in the document]
Comment (LiangquanLi930) [reply ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml:8]: sure, let's me update, and execute conformance again

File: ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml (hunk 1 above)
Comment (LiangquanLi930) [reply ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml:20]: Currently MCE 2.2 is still using the HyperShift CLI, in the future MCE 2.3 will use a different approach, considering the future scenario if 2.2 and 2.3 coexist, So we design it like this now, if we can easily integrate in the future, we will adjust our structure

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

File: ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml (hunk 1 above)
Comment (davidvossel) [reply ci-operator/step-registry/hypershift/mce/kubevirt/conformance/hypershift-mce-kubevirt-conformance-workflow.yaml:21]: That's fine. If MCE 2.3 ends up being largely the same workflow, we can always change this later on.

File: ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        CLUSTER_NAMESPACE=local-cluster-${CLUSTER_NAME}
+        echo "$(date) Creating HyperShift cluster ${CLUSTER_NAME}"
+      
+        # Enable wildcard routes on the management cluster
+        oc patch ingresscontroller -n openshift-ingress-operator default --type=json -p \
+          '[{ "op": "add", "path": "/spec/routeAdmission", "value": {wildcardPolicy: "WildcardsAllowed"}}]'
+
+
+        echo "$(date) Creating HyperShift cluster ${CLUSTER_NAME}"
+        /usr/bin/hypershift create cluster kubevirt \
```
Comment (davidvossel) [inline ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml:23]: I think we want to use the hypershift cli tool shipped by MCE. RIght now (as this is currently written), we're using the one shipped by the latest hypershift-operator for the OCP version being tested on. That's not exactly the same thing. 

To use the hypershift cli tool that the MCE operator ships, i think we can download it from the cluster itself. The url can be found in the yaml of the consoleclidownload resource. An example of how to find that is below.

```oc get ConsoleCLIDownload hypershift-cli-download -o yaml```

Comment (LiangquanLi930) [reply ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml:23]:
```
➜   oc get csv  -A         
NAMESPACE                              NAME                         DISPLAY                              VERSION   REPLACES   PHASE
multicluster-engine                    multicluster-engine.v2.2.0   multicluster engine for Kubernetes   2.2.0                Succeeded
openshift-operator-lifecycle-manager   packageserver                Package Server                       0.19.0               Succeeded
➜  oc get ConsoleCLIDownload -A
NAME                  DISPLAY NAME                                                    AGE
helm-download-links   helm - Helm 3 CLI                                               2023-02-01T10:30:24Z
oc-cli-downloads      oc - OpenShift Command Line Interface (CLI)                     2023-02-01T10:39:03Z
odo-cli-downloads     odo - Developer-focused CLI for OpenShift (Community Support)   2023-02-01T10:39:03Z
```
I did not find the cr related to hypershift-cli, @rokej Could you help to answer to it? thanks

Comment (rokej) [reply ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml:23]: You need to enable the `hypershift-preview` feature in mce to get the hypershift CLI download.

Comment (rokej) [reply ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml:23]: https://github.com/stolostron/hypershift-addon-operator/blob/main/docs/provision_hosted_cluster_on_mce_local_cluster.md#enabling-the-hosted-control-plane-feature

Comment (LiangquanLi930) [reply ci-operator/step-registry/hypershift/mce/kubevirt/create/2.2/hypershift-mce-kubevirt-create-2.2-chain.yaml:23]: I ignored it, thank you for your reminder!

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

**Comment by LiangquanLi930** [comment]
some problem in 4.13 
```
oc get pod -n local-cluster-kv-nested | grep virt-launcher | wc -l
       0

```
But it should not affect this PR merge. no relation with PR

**Comment by rokej** [comment]
It looks good to me. Thank you @LiangquanLi930

**Review by rokej — APPROVED** [review]

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-main-periodics-4.13-conformance-mce-kubevirt

**Review by davidvossel — COMMENTED** [review]
In general, looks great. I have two comments here.

* 4.12 looks like it's in pretty good shape, but the conformance tests are not all passing. Has that been looked into at all to see if we can at least get the periodic reliably passing?

* 4.13 looks permanently broken. I dug around a bit and it looks like this is related to PSA. I'll start a discussion thread with the right people to undestand what's going on here and copy you on it. Below is the log message that occurs when the vmi pod is attempted to be created.

```{"component":"virt-controller","level":"info","msg":"reenqueuing VirtualMachineInstance local-cluster-kv-nested/kv-nested-wwlpw","pos":"vmi.go:273","reason":"failed to create virtual machine pod: pods \"virt-launcher-kv-nested-wwlpw-\" is forbidden: unable to validate against any security context constraint: [provider \"anyuid\": Forbidden: not usable by user or serviceaccount, pod.metadata.annotations[seccomp.security.alpha.kubernetes.io/pod]: Forbidden: localhost/kubevirt/kubevirt.json is not an allowed seccomp profile. Valid values are [runtime/default], spec.containers[0].securityContext.runAsUser: Invalid value: 107: must be in the ranges: [1000790000, 1000799999], pod.metadata.annotations[container.seccomp.security.alpha.kubernetes.io/compute]: Forbidden: localhost/kubevirt/kubevirt.json is not an allowed seccomp profile. Valid values are [runtime/default], provider \"restricted\": Forbidden: not usable by user or serviceaccount, provider \"containerized-data-importer\": Forbidden: not usable by user or serviceaccount, provider \"nonroot-v2\": Forbidden: not usable by user or serviceaccount, provider \"nonroot\": Forbidden: not usable by user or serviceaccount, provider \"hostmount-anyuid\": Forbidden: not usable by user or serviceaccount, pod.metadata.annotations[seccomp.security.alpha.kubernetes.io/pod]: Forbidden: localhost/kubevirt/kubevirt.json is not an allowed seccomp profile. Valid values are [runtime/default unconfined], pod.metadata.annotations[container.seccomp.security.alpha.kubernetes.io/compute]: Forbidden: localhost/kubevirt/kubevirt.json is not an allowed seccomp profile. Valid values are [runtime/default unconfined], provider \"machine-api-termination-handler\": Forbidden: not usable by user or serviceaccount, provider \"bridge-marker\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork-v2\": Forbidden: not usable by user or serviceaccount, provider \"hostnetwork\": Forbidden: not usable by user or serviceaccount, provider \"hostaccess\": Forbidden: not usable by user or serviceaccount, provider \"linux-bridge\": Forbidden: not usable by user or serviceaccount, provider \"kubevirt-handler\": Forbidden: not usable by user or serviceaccount, provider \"node-exporter\": Forbidden: not usable by user or serviceaccount, provider \"privileged\": Forbidden: not usable by user or serviceaccount]","timestamp":"2023-02-08T18:30:55.728131Z"}```

**Comment by LiangquanLi930** [comment]
As 4.13 still have issues, and we need more time to discuss and confirm. I will remove 4.13 changes from this PR right now. I will raise a new one when confirmed.

**Comment by LiangquanLi930** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.12-periodics-4.12-conformance-mce-kubevirt

**Comment by davidvossel** [comment]
fyi, the cnv fix for 4.13 depends on this upstream PR getting merged. https://github.com/kubevirt/kubevirt/pull/9209

**Review by davidvossel — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**